noop/shared/src/main/scala/org/typelevel/log4cats/noop/NoOpLogger.scala

```diff
@@ -28,44 +28,83 @@ object NoOpLogger {
-        void(msg)
-      @inline override def warn(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
-        void(msg)
-      @inline override def error(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
-        void(msg)
+  private def lazyImpl[F[_]: Applicative] = new SelfAwareStructuredLogger[F] {
+
+    val no: F[Boolean] = Applicative[F].pure(false)
+    val unit: F[Unit] = Applicative[F].pure(())
+
+    @inline override def isTraceEnabled: F[Boolean] = no
+    @inline override def isDebugEnabled: F[Boolean] = no
+    @inline override def isInfoEnabled: F[Boolean] = no
+    @inline override def isWarnEnabled: F[Boolean] = no
+    @inline override def isErrorEnabled: F[Boolean] = no
+    @inline override def trace(t: Throwable)(msg: => String): F[Unit] = unit
+    @inline override def trace(msg: => String): F[Unit] = unit
+    @inline override def trace(ctx: Map[String, String])(msg: => String): F[Unit] = unit
+    @inline override def debug(t: Throwable)(msg: => String): F[Unit] = unit
+    @inline override def debug(msg: => String): F[Unit] = unit
+    @inline override def debug(ctx: Map[String, String])(msg: => String): F[Unit] = unit
+    @inline override def info(t: Throwable)(msg: => String): F[Unit] = unit
+    @inline override def info(msg: => String): F[Unit] = unit
+    @inline override def info(ctx: Map[String, String])(msg: => String): F[Unit] = unit
+    @inline override def warn(t: Throwable)(msg: => String): F[Unit] = unit
+    @inline override def warn(msg: => String): F[Unit] = unit
+    @inline override def warn(ctx: Map[String, String])(msg: => String): F[Unit] = unit
+    @inline override def error(t: Throwable)(msg: => String): F[Unit] = unit
+    @inline override def error(msg: => String): F[Unit] = unit
+    @inline override def error(ctx: Map[String, String])(msg: => String): F[Unit] = unit
+    @inline override def trace(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
+      unit
+    @inline override def debug(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
+      unit
+    @inline override def info(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
+      unit
+    @inline override def warn(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
+      unit
+    @inline override def error(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
+      unit
+  }
+
+  private def stricImpl[F[_]: Applicative] = new SelfAwareStructuredLogger[F] {
+
+    val no: F[Boolean] = Applicative[F].pure(false)
+    def void(arg: => Any): F[Unit] = Applicative[F].pure {
+      val _ = arg
+      ()
     }
+
+    @inline override def isTraceEnabled: F[Boolean] = no
+    @inline override def isDebugEnabled: F[Boolean] = no
+    @inline override def isInfoEnabled: F[Boolean] = no
+    @inline override def isWarnEnabled: F[Boolean] = no
+    @inline override def isErrorEnabled: F[Boolean] = no
```

Comment:
The main use case of the `strictImpl` is to be able to exercise a much as possible code paths under test. So I'm wondering that maybe we should say yes instead here, in case the client code tests this before doing stuff

Reply:
That sounds reasonable to me.